### PR TITLE
Implement IHttpResetFeature in Kestrel

### PR DESCRIPTION
**Pull request by Tratcher**
#10886 We already had the ResetAndAbort functionality, we just needed a feature to expose it.

### CODE REVIEW

File: src/Servers/Kestrel/Core/src/Internal/Http2/Http2Stream.FeatureCollection.cs [hunk 1]
```diff
@@ -73,5 +75,11 @@ async Task IHttpResponseCompletionFeature.CompleteAsync()
                 await ProduceEnd();
             }
         }
+
+        void IHttpResetFeature.Reset(int errorCode)
+        {
+            var abortReason = new ConnectionAbortedException(CoreStrings.FormatHttp2StreamResetByApplication((Http2ErrorCode)errorCode));
```
Comment (JamesNK) [inline src/Servers/Kestrel/Core/src/Internal/Http2/Http2Stream.FeatureCollection.cs]: Do we want to validate the errorCode is a valid Http2ErrorCode?

File: src/Http/Http.Features/src/IHttpResetFeature.cs [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+namespace Microsoft.AspNetCore.Http.Features
+{
+    /// <summary>
+    /// Used to send reset messages for protocols that support them such as HTTP/2 or HTTP/3.
+    /// </summary>
+    public interface IHttpResetFeature
+    {
+
```
Comment (JamesNK) [inline src/Http/Http.Features/src/IHttpResetFeature.cs]: nit: liney

File: src/Http/Http.Features/src/IHttpResetFeature.cs [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+        /// <summary>
+        /// Send a reset message with the given error code. The request will be considered aborted.
+        /// </summary>
+        /// <param name="errorCode"></param>
```
Comment (JamesNK) [inline src/Http/Http.Features/src/IHttpResetFeature.cs]: Docs?

File: src/Servers/Kestrel/Core/src/Internal/Http2/Http2Stream.FeatureCollection.cs (hunk 1 above)
Comment (JamesNK) [inline src/Servers/Kestrel/Core/src/Internal/Http2/Http2Stream.FeatureCollection.cs]: What happens if Reset is called multiple times for a request?

What happens if Reset is called after the request has already been aborted? (e.g. the client sends RST_STREAM)

Comment (Tratcher) [reply src/Servers/Kestrel/Core/src/Internal/Http2/Http2Stream.FeatureCollection.cs]: No, they can add new ones at any time. Also the server (and in most cases the client) does not behave any differently based on the value.

Comment (Tratcher) [reply src/Servers/Kestrel/Core/src/Internal/Http2/Http2Stream.FeatureCollection.cs]: ResetAndAbort is guarded against multiple invocations.
https://github.com/aspnet/AspNetCore/blob/9c4faf86e5534dba490dc0e2f634ae291aaa2cfd/src/Servers/Kestrel/Core/src/Internal/Http2/Http2Stream.cs#L452-L460

**Comment by JamesNK** [comment]
Should this raise the RequestAborted cancellation token in TestServer when called?

**Comment by Tratcher** [comment]
@JamesNK if we implemented it in TestServer... It's a specialized Abort.
That said, this is a weird one to implement in TestServer, there's no way to surface the reset to the client.

**Comment by JamesNK** [comment]
Abort (and CompleteAsync) in TestServer should probably return the HttpResponseMessage to the client immediately instead of waiting for the request delegate to complete. That would be a non-minor change.

I think at least we can make it raise the CT like Abort does - https://github.com/aspnet/AspNetCore/pull/9975

**Comment by Tratcher** [comment]
TestServer should already be returning the response message when you write/flush the response body. Making CompleteAsync work on top of that should be trivial, it just needs to additionally close the response body.

Abort on TestServer should throw on the client, not return a response message.

I'd be tempted to only expose IHttpResetFeature in TestServer for request that set the request protocol to HTTP/2.

IHttpResetFeature in TestServer behaving the same as Abort seems like a good start.

I can do a follow-up PR for TestServer CompleteAsync and Reset.

**Review by JamesNK — APPROVED** [review]

**Review by halter73 — APPROVED** [review]

**Pull request merged**